GpsDataAnalyzer/fileparser/fileparser.py

```diff
@@ -46,14 +50,115 @@ def parse_file(self, filename) -> GpsDataSet:
+            logger.info('Parse csv file ' + filename)
             return self.parse_csv(filename)
         else:
-            print("Invalid file type!")
+            logger.debug('Invalid file type.')
             return None
 
 
+    def parse_time(self, timestr) -> datetime:
+        """
+        Parse the xml timestamp and return a new datetime object
+
+        Args:
+          timestr: string of the timestamp extracted from xml file
+
+        Returns:
+          A datetime object with timezone
+        """
+        #Replace timezone Z with '' since python datetime parser could not handler %Z
+        timestr = timestr.replace('Z', '', 1)
+        #Parse time in the format "%Y-%m-%dT%H:%M:%S.%f"
+        time = datetime.strptime(timestr, "%Y-%m-%dT%H:%M:%S.%f")
+        #Set timezone back
+        tz_time = time.replace(tzinfo=timezone.utc)
+
+        return tz_time
+
+
+    def calculate_distance(self, location1, location2) -> float:
+        """
+        Calculate 3D distance (including altitude) between two location points with ellipsoidal earth model
+
+        Args:
+          location1: tuple of (latitude, longitude) as floats in Decimal Degrees
+          location2: tuple of (latitude, longitude) as floats in Decimal Degrees
+
+        Returns:
+           A float in meters of the distance between two location points
+        """
+        return geodesic(location1, location2).meters
+
+
+    def parse_xml_metadata(self, root, prefix) -> GpsMetaData:
+        """
+        Helper function to parse metadata information in xml file
+        """
+        #Get metadata information
+        metadata = root.find(prefix + 'metadata')
+        # Parse start timestamp
+        starttimestr = metadata.find(prefix + 'time').text
+        tz_starttime = self.parse_time(starttimestr)
+
+        device = metadata.find(prefix + 'device').text
+        identifier = metadata.find(prefix + 'id').text
+        manufacturer = metadata.find(prefix + 'manufacturer').text
+        model = metadata.find(prefix + 'model').text
+
+        # Parse end timestamp
+        endtimestr = root.find(prefix + 'time').text
+        tz_endtime = self.parse_time(endtimestr)
+        
+        # Create the GpsMetaData
+        return GpsMetaData(device, identifier, manufacturer, model, tz_starttime, tz_endtime)
+
+
+    def parse_xml_trkpts(self, root, prefix) -> []:
+        """
+        Helper function to parse trkpts in xml file
+        """
+        xml_gpsdatalist = []
+
+        # Find the start location
+        first_trkpt = root.find(prefix + 'trk').find(prefix + 'trkseg')[0]
+        prev_location = (first_trkpt.get('lat'), first_trkpt.get('lon'))
+
+        # Get every track point information
+        for trkpt in root.find(prefix + 'trk').find(prefix + 'trkseg'):
+            # Get the latitude and longitude
+            lat = float(trkpt.get('lat'))
+            lon = float(trkpt.get('lon'))
+
+            # Get the altitude if has
+            ele = None
```

Comment:
does ele = elevation?
the comment talks about altitude so perhaps we should name this variable 'alt' to be consistent?

Reply:
Will do.